astropy/stats/sigma_clipping.py

```diff
@@ -101,281 +192,428 @@ class SigmaClip:
+                if HAS_BOTTLENECK:
+                    cenfunc = _nanmean
+                else:
+                    cenfunc = np.nanmean  # pragma: no cover
+
+            else:
+                raise ValueError('{} is an invalid cenfunc.'.format(cenfunc))
+
+        return cenfunc
+
+    def _parse_stdfunc(self, stdfunc):
+        if isinstance(stdfunc, str):
+            if stdfunc != 'std':
+                raise ValueError('{} is an invalid stdfunc.'.format(stdfunc))
+
+            if HAS_BOTTLENECK:
+                stdfunc = _nanstd
+            else:
+                stdfunc = np.nanstd  # pragma: no cover
+
+        return stdfunc
+
+    def _compute_bounds(self, data, axis=None):
+        # ignore RuntimeWarning if the array (or along an axis) has only
+        # NaNs
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore", category=RuntimeWarning)
+            self._max_value = self.cenfunc(data, axis=axis)
+            std = self.stdfunc(data, axis=axis)
+            self._min_value = self._max_value - (std * self.sigma_lower)
+            self._max_value += std * self.sigma_upper
+
+    def _sigmaclip_noaxis(self, data, masked=True, copy=True):
+        filtered_data = data.ravel()
+
+        # remove masked values and convert to ndarray
+        if isinstance(filtered_data, np.ma.MaskedArray):
+            filtered_data = filtered_data.data[~filtered_data.mask]
+
+        # remove invalid values
+        good_mask = np.isfinite(filtered_data)
+        if np.any(~good_mask):
+            filtered_data = filtered_data[good_mask]
+            warnings.warn('Input data contains invalid values (NaNs or '
+                          'infs), which were automatically clipped.',
+                          AstropyUserWarning)
 
-            min_value = min_value.reshape(mshape)
-            max_value = max_value.reshape(mshape)
+        nchanged = 1
+        iteration = 0
+        while nchanged != 0 and (iteration < self.maxiters):
+            iteration += 1
+            size = filtered_data.size
+            self._compute_bounds(filtered_data, axis=None)
+            filtered_data = filtered_data[(filtered_data >= self._min_value) &
+                                          (filtered_data <= self._max_value)]
+            nchanged = size - filtered_data.size
+
+        self._niterations = iteration
+
+        if masked:
+            # return a masked array
+            data = np.ma.masked_invalid(data, copy=copy)
+
+            # update the mask in place, ignoring RuntimeWarnings for
+            # comparisons with NaN data values
+            with np.errstate(invalid='ignore'):
+                data.mask |= np.logical_or(data < self._min_value,
+                                           data > self._max_value)
+
+            return data
+        else:
+            # return the truncated array with the mix/max clipping thresholds
+            return filtered_data, self._min_value, self._max_value
```

Comment:
Why are the min/max returned in the ndarray case but not the masked array case? I think it might be nicer to have a separate option to determine if the min/max are returned or not that is independent of the ``masked`` option.

Reply:
I added a `return_bounds` bool option.   The default is `False` to preserve backwards compatibility.